src/engine/Match_tainting_mode.ml

```diff
@@ -383,33 +373,46 @@ let any_is_in_sources_matches rule any matches =
               })
          else None)
 
+(* Builds a table to quickly look up for propagator matches, taking advantage of
+ * the requirement that propagators must be exact matches.
+ *
+ * OBS: Previously we allowed these matches if they had an overlap of >0.99, but
+ *   to be honest that was arbitrary and fragile (the overlap largely depends on
+ *   the amount of text being matched). Typically propagators from/to match
+ *   l-values and those typically we can count as being 100% perfect matches.
+ *   If this causes problems we can always roll back, but all tests still work.
+ *)
+let propagators_table_of_matches rule matches =
+  let mk_match prop var kind r =
+    let spec_pm = RM.range_to_pattern_match_adjusted rule prop.rwm in
+    let spec : D.a_propagator = { kind; prop = prop.spec; var } in
+    {
+      Taint_smatch.spec;
+      spec_id = prop.spec.propagator_id;
+      spec_pm;
+      range = r;
+      overlap = 1.0;
+    }
+  in
+  let tbl = Hashtbl.create 100 in
```

Comment:
How large does this table get on larger repos?

Reply:
Also, how does the compare for this work? Is it using polymorphic compare?

Reply:
We've been burned before by polymorphic compare so I'd appreciate the dedicated hash table :D If we were previously allocating a list every time I'm fine with you skipping the concrete numbers, that reasoning convinces me.